content/en/docs/components/pipelines/sdk/component-development.md

```diff
@@ -1,85 +1,93 @@
-Note that the Pipelines system cannot provide consistency and reproducibility guarantees for the data outside of its control.
-
-### Consuming data
-
-There are two main ways a command-line program usually consumes data:
-
-* Small pieces of data are usually passed as command-line arguments: `program.py --param1 100`.
-* Bigger data or binary data is usually stored in files and then the file paths are passed to the program: `program.py --data1 /inputs/data1.txt`. The system needs to be aware about the need to put some data into files and pass their paths to the program.
-
-## Writing the program code
-
-This section describes an example program that has two inputs (for small and
-large pieces of data) and one output. Although the programming language in this example
-is Python 3, you can build your component in any language.
+```python
+$ pip3 install kfp --upgrade
+```
 
-### program.py
+For more information about the Kubeflow Pipelines SDK, see the [SDK reference guide][sdk-ref].
+
+[sdk-ref]: https://kubeflow-pipelines.readthedocs.io/en/latest/index.html
+
+## Understanding pipeline components
+
+Pipeline components are self-contained sets of code that perform one step in
+your ML workflow, such as preprocessing data or training a model. To create a
+component, you must _build the component's implementation_ and _define the
+component specification_.
+
+Your component's implementation includes the component's executable code and
+the Docker container image that the code runs in.  [Learn more about designing
+a pipeline component](#design).
+
+Once you have built your component's implementation, you can define your
+component's interface as a component specification. A component specification
+defines:
+
+*   The component's inputs and outputs.
+*   The container image that your component's code runs in, the command to use
+    to run your component's code, and the command-line arguments to pass to
+    your component's code.
+*   The component's metadata, such as the name and description.
+
+[Learn more about creating a component specification](#component-spec).
+
+If your component's code is implemented as a Python function, use the
+Kubeflow Pipelines SDK to package your function as a component. [Learn more
+about building Python function-based components][python-function-components].
+
+[python-function-components]: https://www.kubeflow.org/docs/pipelines/sdk/python-function-components/
+
+<a name="design"></a>
+## Designing a pipeline component
+
+When Kubeflow Pipelines executes a component, a container image is started in a
+Kubernetes Pod and your component's inputs are passed in as command-line
+arguments. You can pass small inputs, such as strings and numbers, by value.
+Larger inputs, such as CSV data, must be passed as paths to files. When your
+component has finished, the component's outputs are returned as files.
+
+When you design your component's code, consider the following:
```

Comment:
This might be a good place to clarify that KFP components have two broad categories of I/O:
* parameters (ints, floats, small strings)
* artifacts (file locations or other URIs (e.g. API resources) pointing to your Dataset, Model, Metrics)

Reply:
Is this a distinction in how inputs are passed in V2? Or, is this a distinction in how MLMD stores data in v1?

Reply:
It's a distinction in how data is passed in v2 actually. Having the wording introduced here still makes sense, rather than "small data" vs "big data". WDYT?

Reply:
Since it is a v2 distinction, I would recommend that we leave it out and update the doc  as part of the v2 changes. That way we don't mix the recommendations for the v1 and v2 DSL.

Reply:
>we don't mix the recommendations for the v1 and v2 DSL.

+1 I think having a clear separation between v1 and v2 docs prevents confusion.

>KFP components have two broad categories of I/O

The KFP components by their design do not discriminate input and output kinds. There is a single kind of output, "output". Some execution systems might decide to discriminate, based on type. Some might decide not to. TFX for example, has only one kind of output as well. Since the doc talks about building components (not about the execution engines), I'm not sure we should confuse the users with concepts that are implementation details of some of the execution engines.
On the other hand, the data size limitation is relevant since it's inherent to all command-line programs: There is an OS-imposed size limit for the size of command-line argument. There is also a limitation that command-line arguments cannot contain null bytes (`\x00`) which leads to the suggestion not to consume binary data by value.